uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Import/Parser/BaseTest.php

```diff
@@ -0,0 +1,288 @@
+    $result = $leaveRequestCommentService->get(['leave_request_id' => $leaveRequest->id, 'sequential' => 1]);
+    $this->assertEquals($result['values'][0]['text'], $row1['comments']);
+    $this->assertEquals($result['values'][0]['contact_id'], $row1['contact_id']);
+    $commentDate = new DateTime('2016-01-01');
+    $expectedCommentDate = new DateTime($result['values'][0]['created_at']);
+    $this->assertEquals($expectedCommentDate, $commentDate);
+  }
+
+  /**
+   * @expectedException CRM_HRLeaveAndAbsences_Exception_InvalidLeaveRequestException
+   * @expectedExceptionMessage Contact does not have period entitlement for the absence type
+   */
+  public function testValidateLeaveParamsThrowsExceptionWhenContactHasNoPeriodEntitlementForTheAbsenceType() {
+    AbsencePeriodFabricator::fabricate([
+      'start_date' => CRM_Utils_Date::processDate('2016-01-01'),
+      'end_date'   => CRM_Utils_Date::processDate('2016-12-31'),
+    ]);
+
+    $contactId = 1;
+    $row1 = [
+      'contact_id' => $contactId,
+      'absence_id' => 1,
+      'type_id' => $this->absenceType->id,
+      'absence_date' => '2016-01-01',
+      'qty' => 2,
+      'start_date' => '2016-01-01',
+      'end_date' => '2016-01-01',
+      'total_qty' => 3,
+      'status_id' => 1,
+      'request_type' => LeaveRequest::REQUEST_TYPE_LEAVE,
+    ];
+
+    $fields = array_keys($row1);
+    $importObject = $this->getImportObject($fields);
+    $importObject->validateLeaveParams($row1);
+  }
+
+  public function testValidateLeaveParamsThrowsExceptionWhenBalanceIsGreaterThanEntitlementBalanceAndAllowOveruseFalse() {
+    $period = AbsencePeriodFabricator::fabricate([
+      'start_date' => CRM_Utils_Date::processDate('2016-01-01'),
+      'end_date'   => CRM_Utils_Date::processDate('2016-12-31'),
+    ]);
+
+    $absenceType = AbsenceTypeFabricator::fabricate([
+      'title' => 'Type 1',
+      'allow_overuse' => 0
+    ]);
+
+    $periodEntitlement = LeavePeriodEntitlementFabricator::fabricate([
+      'type_id' => $absenceType->id,
+      'contact_id' => 1,
+      'period_id' => $period->id
+    ]);
+
+    $entitlementBalanceChange = 3;
+    $this->createLeaveBalanceChange($periodEntitlement->id, $entitlementBalanceChange);
+
+    HRJobContractFabricator::fabricate(
+      ['contact_id' => $periodEntitlement->contact_id],
+      ['period_start_date' => '2016-01-01']
+    );
+
+    WorkPatternFabricator::fabricateWithA40HourWorkWeek(['is_default' => 1]);
+
+    //This will create a balance change of 4 since the value for total_qty
+    //is 4
+    $row1 = [
+      'contact_id' => $periodEntitlement->contact_id,
+      'absence_id' => 1,
+      'type_id' => $absenceType->id,
+      'absence_date' => '2016-01-01',
+      'qty' => 4,
+      'start_date' => '2016-01-01',
+      'end_date' => '2016-01-01',
+      'total_qty' => 4,
+      'status_id' => 1,
+      'request_type' => LeaveRequest::REQUEST_TYPE_LEAVE,
+    ];
+
+    $this->setExpectedException(
```

Comment:
What about moving this to right before the `validateLeaveParams()` call to make it clear that we're expecting this method to throw the exception?